# 3.3: Notebook-Exercise Solutions-String Types

## Exercise 1

Use each of the strings introduced above in a print command and consider the output. For example, does the format of the string change the output? What happens if the string is longer than the line where it is printed?

### Solution 1

The strings introduced in the reading are:

In [5]:
print('This is a string.')
print("So is this!")
print("This string contains a quote symbol, but it's not the end of the string.")
print(""" Triple quoted strings provide 

    an easy way of incorporating 

    chunks of formatted material into a program file.""")
print("Special characters like 'newlines' \n and 'tabs' \t can be included by use of backslashes")

This is a string.
So is this!
This string contains a quote symbol, but it's not the end of the string.
 Triple quoted strings provide 

    an easy way of incorporating 

    chunks of formatted material into a program file.
Special characters like 'newlines' 
 and 'tabs' 	 can be included by use of backslashes


The formatting (i.e. the line breaks and spacing) of the fourth string is captured by the print statement, and the new line and tab characters add additional spacing to the final example. If none of the above examples are longer than where they're printed, you might try your own example. Using this text:

In [ ]:
print("The formatting (i.e. the line breaks and spacing) of the fourth string is captured by the print statement, and the new line and tab characters add additional spacing to the final example. If none of the above examples are longer than where they're printed, you might try your own example. Using this text:")

Which (should) wrap the text, rather than continuing to print off screen.

## Exercise 2

In Python, which is smaller "post office" or "post-haste"?

### Solution 2

In [1]:
"post office" < "post-haste"

True

So Python considers a space less than a hyphen.

## Exercise 3
1. Are startswith and endswith case sensitive?
2. Use split and len to count how many words there are in the sentence "You must be the change you wish to see in the world". Hint: you can use the output of split as the argument of len.
3. Use split and join to remove the hyphen characters in the sentence "I find the use of hyphens counter-intuitive and not straight-forward".

### Solution 3

1. `startswith` and `endswith` are case sensitive:

In [ ]:
'fundamentals of programming'.startswith('Fun')

In [ ]:
'FUNdamentals of programming'.startswith('fun')

In [ ]:
'slaughter'.endswith('LAughter')

In [ ]:
'sLaughter'.endswith('LAughter')

2. There are 12 words in the sentence "You must be the change you wish to see in the world", which we can do by using the method `split` to turn the string into a list of words, then use `len` to count the number of elements in the list:  

In [ ]:
len("You must be the change you wish to see in the world".split())

3. We can split the sentence "I find the use of hyphens counter-intuitive and not straight-forward" up at the hyphens, then join using an empty string:

In [ ]:
' '.join("I find the use of hyphens counter-intuitive and not straight-forward".split('-'))

## Exercise 4

1. Does the order in which the options are specified matter?
2. Does the total width specified for a float include the space for the decimal place?
3. What happens if the total width specified for a float is less than the width of the float?
4. The b option converts a decimal integer to binary, but does the d option convert binary to decimal?
5. There are additional formatting options to those described above. What do you think the formatting options do in f'{12345.678:*<+15,.4f}'?Would the same specification work with `.format()`?
6. Rewrite the following f-string using `.format()`: `f"The value of x is {x} and x squared is {x**2}"`
7. Rewrite the following `.format()` call using an f-string: `"Pi is approximately {:.4f}".format(3.14159)`


### Solution 4

1. Yes, order matters, for example, we can't put the left align character after specifying fixed point notation:

In [ ]:
# This produces an error
f'{1/7:20.12f<}'

2. Yes, the width includes the decimal point. In the example below we fill empty space with 0's and the total width is 16 characters:

In [ ]:
f'{1/7:016.12f}'

3. Python is quite forgiving, so it will increase the width to ensure the number of decimal places is correct. In the example below, the width is specified as 5, but the string returned has 7 characters width.

In [ ]:
f'{100/7:5.4f}'

4. A number with just zeros and ones is not cast into a binary number, since it could also be a decimal number:

In [ ]:
f'{10011010010:d}'

However, a number can be specified as binary by preceeding it by `0b`, and this will be converted to decimal:

In [ ]:
f'{0b10011010010:d}'

In [ ]:
f'{1234:b}'

5. The code `f'{12345.678:*<+15,.4f}'` uses several formatting options:
   - `*` - Fill character (asterisks fill empty space)
   - `<` - Left align the number
   - `+` - Show sign for both positive and negative numbers
   - `15` - Total width of 15 characters
   - `,` - Use comma as thousands separator
   - `.4f` - Fixed-point notation with 4 decimal places

Let's see the output:

In [2]:
f'{12345.678:*<+15,.4f}'

'+12,345.6780***'

This produces `'+12,345.6780***'` - the number is left-aligned, has a plus sign, uses comma separators, has 4 decimal places, and asterisks fill the remaining space to reach width 15.

Yes, the same specification works with `.format()`:

In [ ]:
"{:*<+15,.4f}".format(12345.678)

Both f-strings and `.format()` use the same formatting specification language, so any format specification that works in one will work in the other.

6. The f-string `f"The value of x is {x} and x squared is {x**2}"` can be rewritten using `.format()` in several ways:

In [3]:
# Using positional arguments
x = 5
"The value of x is {} and x squared is {}".format(x, x**2)

'The value of x is 5 and x squared is 25'

In [ ]:
# Using named arguments (most explicit)
x = 5
"The value of x is {value} and x squared is {squared}".format(value=x, squared=x**2)

7. The `.format()` call `"Pi is approximately {:.4f}".format(3.14159)` can be rewritten as an f-string:

In [ ]:
pi = 3.14159
f"Pi is approximately {pi:.4f}"

Or if we don't want to define a variable first, we can use the expression directly:

In [ ]:
f"Pi is approximately {3.14159:.4f}"